### PR TITLE
Fix generation of next gen images with first settings save of avif option

### LINKED ISSUE #825: 1st activation of avif won't run automatically in the background
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version => commit 324c4fa030cb3197aef071d09845184f0819fe9f
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
When we enable AVIF for the 1st time after fresh activation, AVIF won't be automatically generated (reactivate avif will make the creation works automatically)

**To Reproduce**
1. install and activate imagify
2. upload some images (small, medium, big)
3. activate avif => (notice that bulk notice isnot displayed)
4. check media library => generate next-gen link displayed (nothing generated in background)

**Expected behavior**
Automatic creation of avif works even when activating the option for the 1st time

**Screenshots**
Video and sample images
[screen-capture (84).webm](https://github.com/wp-media/imagify-plugin/assets/76941962/9ab3200f-5ffd-4390-b21b-d07096072e49)
download (33)
file_example_JPG_100kB (1)
mountain

**Additional context**
- Reactivate AVIF will work

**Acceptance Criteria (for WP Media team use only)**
Clear instructions for developers, to be added before the grooming

### PR DESCRIPTION
**Pull request by wordpressfan**
# Description

Run next gen generation for the first settings save and add few comments in code to explain what is going on there.

Fixes #825

### User documentation

This will trigger generating next generation images when enabling the avif option and rewrite rules.

### Technical documentation

*Explain how this code works. Diagram & drawings are welcomed.*

## Type of change
*Delete options that are not relevant.*

- [x] Bug fix (non-breaking change which fixes an issue).

## New dependencies
*List any new dependencies that are required for this change.*

## Risks
*List possible performance & security issues or risks, and explain how they have been mitigated.*

# Checklists

## Feature validation

- [x] I validated all the Acceptance Criteria. If possible, provide sreenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [ ] I implemented built-in tests to cover the new/changed code.

## Documentation

- [x] I prepared the user documentation for the feature/enhancement and shared it in the PR or the GitHub issue.
- [x] The user documentation covers new/changed entry points (endpoints, WP hooks, configuration files, ...).
- [ ] I prepared the technical documentation if needed, and shared it in the PR or the GitHub issue.

## Code style
- [x] I wrote self-explanatory code about what it does.
- [x] I wrote comments to explain why it does it.
- [x] I named variables and functions explicitely.
- [ ] I protected entry points against unexpected inputs.
- [x] I did not introduce unecessary complexity.
- [x] I listed the introduced external dependencies explicitely on the PR.
- [x] I validated the repo-specific guidelines from CONTRIBUTING.md.

## Observability
- [x] I handled errors when needed.
- [ ] I wrote user-facing messages that are understandable and provide actionable feedbacks.
- [ ] I prepared ways to observe the implemented system (logs, data, etc.).

## Risks
- [ ] I explicitely mentioned performance risks in the PR.
- [ ] I explicitely mentioned security risks in the PR.

### CODE REVIEW

File: classes/Bulk/Bulk.php [hunk 1]
```diff
@@ -598,15 +598,13 @@ public function bulk_get_stats_callback() {
...
 		}
 
-		if ( ! isset( $old_value['convert_to_avif'] ) && ! isset( $value['convert_to_avif'] ) ) {
-			return;
-		}
-
-		if ( isset( $value['convert_to_avif'] ) && ! (bool) $value['convert_to_avif'] ) {
+		if ( empty( $value['convert_to_avif'] ) ) {
+			// new value is disabled, do nothing.
 			return;
```
Comment (remyperona) [inline classes/Bulk/Bulk.php:611]: aren't you missing the case where the old and new value are enabled, we should also bail-out?

Comment (MathieuLamiot) [reply classes/Bulk/Bulk.php:611]: Isn't it caught by this?

```
if ( empty( $old_value['convert_to_avif'] ) === empty( $value['convert_to_avif'] ) ) {
			// Old value = new value so do nothing.
			return;
		}
```
		
Both empty evaluation would be false, so we'd bail out.
It seems to be the expected behavior based on [this](https://wp-media.slack.com/archives/CUT7FLHF1/p1709294384163199?thread_ts=1709286678.369739&cid=CUT7FLHF1)

Comment (remyperona) [reply classes/Bulk/Bulk.php:611]: if the value was previously enabled, and stayed enabled, both index will exist and not be empty

Comment (MathieuLamiot) [reply classes/Bulk/Bulk.php:611]: Yes exactly, so:
empty( $old_value['convert_to_avif'] ) is False
empty( $value['convert_to_avif'] ) is False.
empty( $old_value['convert_to_avif'] ) === empty( $value['convert_to_avif'] ) is True then, no?

So we bail out. Is there a mistake here?

Comment (remyperona) [reply classes/Bulk/Bulk.php:611]: ah right, morning fog I guess, it works

Comment (MathieuLamiot) [reply classes/Bulk/Bulk.php:611]: ☕  😄 
To be fair, this is not really straightforward to read 😬

**Comment by MathieuLamiot** [comment]
Don't block the PR for this, but it could be nice to update the doc string or comment the method here to clearly state:
- the expected behavior (only run when moving the avif conversion for off to on)
- the parameter's expected value (not set / 1)
That could avoid some headache next time we work on this :D

**Comment by Mai-Saad** [comment]
Now AVIF is created for 1st activation. with the following note:
-  on main branch: when deactivate avif, we will automatically generate missing webP for existing image => on PR we won't automatically create webP for existing image in this case => in this case the generate missing next-gen will keep loading but nothing will be generated 
@piotrbak  what do you think? In this case,  either we need to automatically generate webp for existing images when deactivate avif or not automatically clicking generate missing next-gen btn.

**Comment by piotrbak** [comment]
@Mai-Saad This behaviour is acceptable for the initial 2.2

**Comment by piotrbak** [comment]
@Mai-Saad @wordpressfan I can see that the PR is still a draft/in-progress. Where are we with this one?

**Comment by wordpressfan** [comment]
I think we are fine here based on the comments above but I requested a review from @Mai-Saad @Tabrisrp and @MathieuLamiot to add their review.

**Review by remyperona — APPROVED** [review]

**Review by MathieuLamiot — APPROVED** [review]

**Review by Mai-Saad — APPROVED** [review]

**Pull request merged**